net/discovery.go

```diff
@@ -83,50 +96,65 @@ func (d *discovery) stop() error {
 	return d.dht.Close()
 }
 
-func (d *discovery) advertise() {
+func (d *discovery) advertiseLoop() {
 	ttl := initialAdvertisementTimeout
 
-	doAdvertise := func() {
-		log.Debug("advertising in the DHT...")
-		err := d.dht.Bootstrap(d.ctx)
-		if err != nil {
-			log.Warnf("failed to bootstrap DHT: err=%s", err)
-			return
-		}
-
-		for _, provides := range d.provides {
-			ttl, err = d.rd.Advertise(d.ctx, string(provides))
-			if err != nil {
-				log.Debugf("failed to advertise in the DHT: err=%s", err)
-				ttl = tryAdvertiseTimeout
-				return
-			}
-		}
-
-		ttl, err = d.rd.Advertise(d.ctx, "")
-		if err != nil {
-			log.Debugf("failed to advertise in the DHT: err=%s", err)
-			ttl = tryAdvertiseTimeout
-			return
-		}
-
-		ttl = defaultAdvertiseTTL
-	}
+	ttl = d.advertise(ttl)
 
 	for {
 		select {
 		case <-d.advertiseCh:
-			// TODO: check current offers, as this may change (#160)
 			d.provides = []types.ProvidesCoin{types.ProvidesXMR}
-			doAdvertise()
+			ttl = d.advertise(ttl)
 		case <-time.After(ttl):
-			doAdvertise()
+			// the DHT clears provider records (ie. who is advertising what content)
+			// every 24 hours.
+			// so, if we don't have any offers available for 24 hours, then we are
+			// no longer present in the DHT as a provider.
+			// otherwise, we'll be present, but no offers will be sent when peers
+			// query us.
+			offers := d.offerAPI.GetOffers()
+			if len(offers) == 0 {
+				continue
+			}
+
+			ttl = d.advertise(ttl)
 		case <-d.ctx.Done():
 			return
 		}
 	}
 }
 
+// advertise advertises that we provide XMR in the DHT.
+// note: we only advertise that we are an XMR provider, but we don't
+// advertise our specific offers.
+// to find what our offers are, peers need to send us a QueryRequest
+// over the query subprotocol.
```

Comment:
Can we mention what the return value is?  It looks like we return the duration the caller should pause before advertising again.